waku/node/peer_manager/peer_manager.nim

```diff
@@ -633,30 +662,88 @@ proc getNumStreams*(pm: PeerManager, protocol: string): (int, int) =
   return (numStreamsIn, numStreamsOut)
 
 proc pruneInRelayConns(pm: PeerManager, amount: int) {.async.} =
-  let (inRelayPeers, outRelayPeers) = pm.connectedPeers(WakuRelayCodec)
+  if amount <= 0:
+    return
+  
+  let (inRelayPeers, _) = pm.connectedPeers(WakuRelayCodec)
   let connsToPrune = min(amount, inRelayPeers.len)
 
   for p in inRelayPeers[0..<connsToPrune]:
+    trace "Pruning Peer", Peer = $p
     asyncSpawn(pm.switch.disconnect(p))
 
-proc connectToRelayPeers*(pm: PeerManager) {.async.} =
-  let (inRelayPeers, outRelayPeers) = pm.connectedPeers(WakuRelayCodec)
-  let maxConnections = pm.switch.connManager.inSema.size
-  let totalRelayPeers = inRelayPeers.len + outRelayPeers.len
-  let inPeersTarget = maxConnections - pm.outRelayPeersTarget
+proc manageRelayPeers*(pm: PeerManager) {.async.} =
+  if pm.wakuMetadata.shards.len == 0:
+    return
+  
+  var peersToConnect: HashSet[PeerId] # Can't use RemotePeerInfo as they are ref objects
+  var peersToDisconnect: int
+
+  # Get all connected peers for Waku Relay
+  var (inPeers, outPeers) = pm.connectedPeers(WakuRelayCodec)
+
+  # Calculate in/out target number of peers for each shards
+  let inTarget = pm.inRelayPeersTarget div pm.wakuMetadata.shards.len
+  let outTarget = pm.outRelayPeersTarget div pm.wakuMetadata.shards.len
 
-  # TODO: Temporally disabled. Might be causing connection issues
-  #if inRelayPeers.len > pm.inRelayPeersTarget:
-  #  await pm.pruneInRelayConns(inRelayPeers.len - pm.inRelayPeersTarget)
+  for shard in pm.wakuMetadata.shards.items:
+    # Filter out peer not on this shard
+    let connectedInPeers = inPeers.filterIt(
+      pm.peerStore.hasShard(it, uint16(pm.wakuMetadata.clusterId), uint16(shard)))
 
-  if outRelayPeers.len >= pm.outRelayPeersTarget:
+    let connectedOutPeers = outPeers.filterIt(
+      pm.peerStore.hasShard(it, uint16(pm.wakuMetadata.clusterId), uint16(shard)))
+
+    # Calculate the difference between current values and targets
+    let inPeerDiff = connectedInPeers.len - inTarget
+    let outPeerDiff = outTarget - connectedOutPeers.len
+
+    if inPeerDiff > 0:
+      peersToDisconnect += inPeerDiff
+
+    if outPeerDiff <= 0:
+      continue
+
+    # Get all peers for this shard
+    var connectablePeers = pm.peerStore.getPeersByShard(
+      uint16(pm.wakuMetadata.clusterId), uint16(shard))
+  
+    let shardCount = connectablePeers.len
+
+    connectablePeers.keepItIf(
+      not pm.peerStore.isConnected(it.peerId) and
+      pm.canBeConnected(it.peerId))
+
+    let connectableCount = connectablePeers.len
+
+    connectablePeers.keepItIf(it.protocols.contains(WakuRelayCodec))
```

Comment:
Based on comments elsewhere, if we do not populate the `ProtoBook` with the proto flags from the ENR, perhaps we can retrieve the ENR here for non-connected peers here and only consider those advertising Relay as viable?